Client RPC migration

tegefaulkes opened the pull request:
### Description

This PR deals with migrating the clientRPC handlers and code to using the new Agnostic RPC system.

It consists of taking the existing handler code within `client/services` and transplanting it into the new `RPC` system from #249 . 

##### pagination 
Any handlers that stream data need to support the pagination parameters and propagate them. The required parameters are `seek: id`, `number: number`, `order: 'asc'|'desc'`. Most code that streams an output with a generator already support pagination. But in some cases where it is not already supported it needs to be updated. As for the handler parameters, the message should include the following

```ts
type paginationMessage = {
id: string,
order: 'asc' | 'desc',
number: number,
}
```

##### Binary data
Currently there are two cases where sending binary data is applicable. The IDs such as `NodeID`, or `VaultId`. These are converted to an encoded string form just for readability of the message. Using the `NodeIdEncoded` and `VaultIdEncoded` for of the Ids. 

The 2nd case is file contents. Currently these are converted to a binary string when provided within a message using `Buffer.toString('binary')`. This is pretty limiting to small amounts of data so it is subject to change. If we need to send larger files we need to stream it using a raw handler to implement it as a raw binary stream. This has complications such it not supporting any middleware so we need to implement authentication logic withing the raw handler.

##### Standard message structure
To standardise the message structure and types across all messages we are leveraging the typescripts type system. Since it's pretty powerful we can compose messages and message types together effectively. 

For example we have basic type messages that can be composed to make more complex messages

```
type NodeIdMessage = {  
  nodeIdEncoded: NodeIdEncoded;  
};

export type AddressMessage = {
  host: string;
  port: number;
};

type NodeAddressMessage = NodeIdMessage & AddressMessage;

type NodesAddMessage = NodeAddressMessage & {  
  force?: boolean;  
  ping?: boolean;  
};

```

### Issues Fixed

* Related #249 
* Related #249 
* Fixes #514

### Tasks

1. [x] Migrate handlers.
	1. [x] agent
	2. [x] gestalts
	3. [x] identities
	4. [x] keys
	5. [x] nodes
	6. [x] notifications
	7. [x] vaults
2. [x] Migrate tests.
	1. [x] agent
	2. [x] gestalts
	3. [x] identities
	4. [x] keys
	5. [x] nodes
	6. [x] notifications
	7. [x] vaults
4. [x] Update BIN code to use new RPC system.
5. [x] Remove old client GRPC code and tests.

### Final checklist
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

tegefaulkes committed:
feat: migrating handlers to agnostic RPC

- created agent handlers and tests
- adding gestalts handlers
- migrating gestalts handlers
- migrating identities handlers
- adding identities handlers
- adding identities handlers
- adding keys handlers
- migrating nodes handlers
- migrating notifications handlers
- migrating vaults handlers
- migrating vaults handlers

[ci skip]

tegefaulkes committed:
tests: migrating handler tests

- migrating keys tests
- migrating nodes tests
- migrating notifications tests
- migrating notifications tests
- migrating vaults tests

[ci skip]

tegefaulkes committed:
feat: updating client RPC usage to use agnostic RPC

- removing client GRPC and using agnostic RPC
- updating agent bin code
- updating identities bin code
- updating keys bin code
- updating nodes bin code
- migrating secrets bin code
- migrating vaults bin code

[ci skip]

tegefaulkes committed:
fix: fixing failing tests

[ci skip]

CMCDragonkai commented on src/grpc/errors.ts:
```diff
@@ -50,7 +50,7 @@ class ErrorGRPCServerVerification<T> extends ErrorGRPC<T> {
 }
 
-class ErrorPolykeyRemote<T> extends ErrorPolykey<T> {
+class ErrorPolykeyRemoteOLD<T> extends ErrorPolykey<T> {
```
Why does this still exist? Is it for the agent RPC?

tegefaulkes replied:
It's the old GRPC stuff, the `agent` RPC still needs it to function. Stuff like this will be fully removed at the end of the agent migration stage.

CMCDragonkai commented on src/RPC/middleware.ts:
```diff
@@ -8,6 +8,7 @@ import type {
+import { promise } from '../utils';
 const jsonStreamParsers = require('@streamparser/json');
 
 function binaryToJsonMessageStream<T extends JsonRpcMessage>(
```
This function does not have a return type?

tegefaulkes replied:
Added a return type.

CMCDragonkai commented on src/RPC/middleware.ts:
```diff
@@ -8,6 +8,7 @@ import type {
 import * as rpcErrors from './errors';
 import * as rpcUtils from './utils';
+import { promise } from '../utils';
 const jsonStreamParsers = require('@streamparser/json');
```
Do not use `require`. Use `import` or `import()`. This should be deprecated. We only use it in very specific set of cases.

tegefaulkes replied:
Import is acting weird with this library. when I use the import semantics or `import()` I get the following error.  `TS2307: Cannot find module '@streamparser/json' or its corresponding type declarations.`

I think it's using module exports. We have problems importing that right?

CMCDragonkai replied:
You can't use `import()` without top level await or being in Async context.

CMCDragonkai replied:
But you should ask upstream and check what you can do

tegefaulkes replied:
It's not an upstream issue. If I asked they'd likely just point to the first line of their example `import { JSONParser } from '@streamparser/json';`. It's more likely a problem with our configuration, I'll look into it some more.

CMCDragonkai replied:
Why can't we import like that?

tegefaulkes replied:
Because when I do I get `TS2307: Cannot find module '@streamparser/json' or its corresponding type declarations.`.

Only using `require()` works for reasons I have not worked out yet.

CMCDragonkai replied:
We don't have this problem with any of the other libraries. Although I think if they are using the new node module system instead of commonjs, then it can be an issue. But we should be working out a way to be able to import both module systems and not be using `require`.

CMCDragonkai replied:
I think you should indicate this upstream. That way you can point out something about our configuration causing import problems, but that doesn't affect other packages we have been using.

CMCDragonkai replied:
Upstream was quite responsive to our previous problems too involving DoS.

tegefaulkes replied:
I'll make  a new issue and an upstream issue for this. It isn't blocking the PR.

tegefaulkes replied:
I have created a new issue for this at #516 .

CMCDragonkai commented on src/RPC/middleware.ts:
```diff
@@ -8,6 +8,7 @@ import type {
 import { TransformStream } from 'stream/web';
```
We need to figure out how to conditionally import this later because in react-native it's in the global scope.

tegefaulkes replied:
Should I make an new issue for this?

CMCDragonkai replied:
No, this is already in TypeScript-Demo-Lib.

CMCDragonkai commented on src/RPC/middleware.ts:
All the functions in this file require docblock documentation explaining what they do.

Furthermore, why are these functions not just part of `src/rpc/utils.ts`?

tegefaulkes replied:
I've added docblocks to these functions now. 

As for why they're not part of `utils.ts` I think I addressed it in one of the other review comments. The're meant to be used when setting up the `RPCServer` or `RPCClient` and not really internally by them. So to keep them from getting lost in the `utils.ts` with the other functions I made them separate. I likely should move both to a `utils` directory though.

I'm resolving this since this discussion can be done in one of the other comments about it.

CMCDragonkai replied:
It may be better to have a `middleware` directory and then separate each middleware by files. That way you can import separately. One may want to separately compose middlewares later.

CMCDragonkai commented on src/RPC/RPCClient.ts:
Hey can you change your directory name to be `src/rpc`. We don't use capitalised directory names in any of our projects...

tegefaulkes replied:
I'll do this, but after all the other changes.

tegefaulkes replied:
This has been done.

CMCDragonkai commented on src/RPC/errors.ts:
```diff
@@ -51,7 +105,8 @@ export {
+  ErrorRpcMissingResponse,
+  ErrorPolykeyRemote,
   ErrorRpcNoMessageError,
   ErrorRpcPlaceholderConnectionError,
```
Change `Rpc` to `RPC`.

tegefaulkes replied:
Updated.

CMCDragonkai commented on src/RPC/errors.ts:
```diff
@@ -51,7 +105,8 @@ export {
+  ErrorRpcMissingResponse,
+  ErrorPolykeyRemote,
   ErrorRpcNoMessageError,
   ErrorRpcPlaceholderConnectionError,
```
What is `ErrorRpcPlaceholderConnectionError`?

tegefaulkes replied:
That was a placeholder error to test some logic. It's a placeholder error for when the underlying streampair errors out due to a connection error or something. In that case we don't want to try and send any error messages back through the stream. I'll have to look it over, remove it and determine if any logic needs to be updated.

tegefaulkes replied:
This should be removed, but I also need to update how we decide to send or not send errors over the RPC. I think the only case so far where we don't send a error message is when we can't due to the connection ending. 

So what I can do here is attempt to send all messages, if sending fails due to a connection failure or whatever then we dispatch the error event contain a compound error of what we tried to send and why it failed.

CMCDragonkai replied:
The usual way to do this is:

1. Handling errors goes to the connection.
2. Connection errors cannot go to the connection, instead it is logged in the actual agent itself, it should not crash the agent
3. On the receiving side, if it gets a handling error, raise it the normal way.
4. If however on the receiving side, the connection is broken (due to keepalive or something else), then raise a special set of errors regarding the reason why the connection is broken

CMCDragonkai replied:
It's important that handling errors AND connection errors does not shutdown the agent. The agent and associated servers should continue to operate.

tegefaulkes replied:
This has been resolved.

CMCDragonkai commented on src/RPC/utils.ts:
```diff
@@ -371,11 +382,14 @@ function clientInputTransformStream<I extends JSONValue>(method: string) {
 }
 
-function clientOutputTransformStream<O extends JSONValue>() {
+function clientOutputTransformStream<O extends JSONValue>(
```
I don't get the distinction between `rpc/utils.ts` and `rpc/middleware.ts`. I think they should be unified.

tegefaulkes replied:
They are kind of the same thing. I created `middleware.ts` to hold the middleware transforms  since they are not really used internally by the RPC but used to configure the middleware when constructing the `RPC`. I can re-combine them, but to me it's neater to have a `middleware.ts` for all your middleware needs.

CMCDragonkai replied:
Like I said it should be either a directory with an `index.ts` or a utils.

CMCDragonkai replied:
Suggest using `middleware/*` where `*` is name of middleware - use camel cased names.

tegefaulkes replied:
Fixed.

CMCDragonkai commented on src/RPC/utils.ts:
```diff
@@ -371,11 +382,14 @@ function clientInputTransformStream<I extends JSONValue>(method: string) {
-function clientOutputTransformStream<O extends JSONValue>() {
+function clientOutputTransformStream<O extends JSONValue>(
+  clientMetadata: ClientMetadata,
+) {
```
No return type?

tegefaulkes replied:
Fixed.

CMCDragonkai commented on src/bin/agent/CommandLockAll.ts:
```diff
@@ -27,7 +26,7 @@ class CommandLockAll extends CommandPolykey {
         this.logger.getChild(binProcessors.processClientOptions.name),
       );
-      const meta = await binProcessors.processAuthentication(
+      const auth = await binProcessors.processAuthentication(
```
What is the type of this `auth`?

tegefaulkes replied:
It used to be a GRPC metadata object. Now it's `{ Authorization?: string }`. I can use it as the `metadata` for the message, or I can combine it with other metadata.

```ts
{
  metadata: auth,
  something: 123,
}
// OR
{
  metadata: {
  some: 'data',
   ...auth,
  }
  some: 'param',
}
```

CMCDragonkai replied:
I think you should use `token` or something similar to how http does for their headers. At the very least should all be lowercase.

tegefaulkes replied:
I'll make it lowercase `authorization?: string`.  It can be a token or the raw password so `authorization` is more appropriate than `token`.

tegefaulkes replied:
Fixed.

CMCDragonkai commented on src/PolykeyClient.ts:
```diff
@@ -1,11 +1,16 @@
-import type { FileSystem, Timer } from './types';
+import type { FileSystem } from './types';
```
Did you delete the `Timer` type in `./types`? Does this mean we no longer use that `Timer`? Have you transitioned everything to using `@matrixai/timer`?

tegefaulkes replied:
Looking at it, it's importing the `@matrixai/timer` now. But looking closely it doesn't seem to be using it. I need to take a closer look at this.

CMCDragonkai replied:
The `Timer` type can be deleted. If nothing is using it.

tegefaulkes replied:
It has been removed.

CMCDragonkai commented on src/PolykeyClient.ts:
```diff
@@ -57,20 +67,26 @@ class PolykeyClient {
+      logger: logger.getChild(WebSocketClient.name),
+      connectionTimeout,
+      pingTimeout,
+    });
```
How can this work?

The `PolykeyClient` needs to be generic to the transports. You cannot hardcode the transport inside here.

For agent RPC, it has to use QUIC. You should factor this out and parameterise the transport.

tegefaulkes replied:
That slipped my mind at the time. At the time I had a problem with providing the `RPCClient` as a parameter, I don't recall exactly why. I'll have the review and fix this because as you mention, this needs to be agnostic to the underlying transport which can be `Quic` or `WebSocket`.

An oversight on my part.

tegefaulkes replied:
This is now fixed, Instead of the transport being internal to the `PolykeyClient`it now takes the `StreamFactory` callback. transport is now created and managed separately.

CMCDragonkai commented on src/PolykeyClient.ts:
```diff
@@ -15,20 +20,23 @@ import config from './config';
-    timer,
+    manifest,
+    connectionTimeout,
+    pingTimeout,
```
How can these work? These are "transport" level properties, should not be part of `PolykeyClient`.

tegefaulkes replied:
This needs to be fixed as per the above comment.

tegefaulkes replied:
This has been fixed.

CMCDragonkai commented on src/client/authenticationMiddleware.ts:
How was this done before when using GRPC. Where did we put our authentication middleware?

tegefaulkes replied:
I don't know about the GRPC specifics of the middleware. If I had to guess it was the interceptors defined in `src/grpc/utils/FlowCountInterceptor.ts`? which is separate to the the utils.ts but part of the utils directory.

CMCDragonkai replied:
That was a hack. That we should get rid of. I'm thinking that either middleware is in its own directory or rolled into utils. No inbetween.

CMCDragonkai commented on src/client/index.ts:
Why delete the `index.ts`?

tegefaulkes replied:
That was the old index.ts, Technically I never created an index.ts when doing this, it's just the old one that was removed. I can create an index.ts.

tegefaulkes replied:
I'll add appropriate `index.ts` files.

tegefaulkes replied:
Fixed.

CMCDragonkai commented on src/client/types.ts:
```diff
@@ -1,14 +1,18 @@
+  } & Partial<{ Authorization: string }>;
+} & Omit<T, 'metadata'>;
+
+export type { RPCRequestParams, RPCResponseResult, NoData };
```
Aren't these generic to `rpc`? Why not have these types in `rpc`?

tegefaulkes replied:
Yes and no? By design the RPC doesn't care about the parameter structure. This is here to enforce the metadata structure for the `client` domain to be used by the middleware, handler and client implementation.

So it's not used at all in the context of the `rpc` domain, but will be used by both the `client` and `agent` domains, they'll share the middleware code as well. So with that maybe this should be moved to a common place between the two? Not sure.

CMCDragonkai replied:
If they are specific to client/agent "services", one should make that clearer by saying `ClientRPCRequestParams` or `AgentRPCRequestParams`. Right now these look like `rpc` domain types.

tegefaulkes replied:
They're getting pretty verbose already. Maybe change them to `ClientRPCReq`, `ClientRPCRes`, `AgentRPCReq` and `AgentRPCRes`?

CMCDragonkai replied:
I don't like shortening type names. Type names can verbose, it's fine.

tegefaulkes replied:
Fixed.

CMCDragonkai commented on src/client/utils.ts:
Are all these functions still being used?

tegefaulkes replied:
Yes, these were transplanted from the old authentication logic for client and updated to the new system.

CMCDragonkai commented on src/client/utils.ts:
Also why change from `client/utils` to `client/utils.ts`? What was the reason behind this?

tegefaulkes replied:
No specific reason. When I was working on this in `clientRPC` before removing the old `client` and renaming `clientRPC` to `client` I just didn't have the `utils.ts` in a `utils` directory. It doesn't seem necessary for a single `utils.ts` file.

I could move `utills.ts` back to a `utils` directory and move `authenticationMiddleware.ts` there as well.

tegefaulkes replied:
fixed.

CMCDragonkai commented on src/nodes/NodeManager.ts:
```diff
@@ -1190,6 +1190,10 @@ class NodeManager {
+    if (seedNodes.length === 0) {
+      logger.debug(`No seed nodes provided, skipping discovery`);
+      return;
+    }
```
What is this about? Why is this here?

tegefaulkes replied:
Technically out of scope of the PR. I added it to remove logger spam during tests. It is technically a fix that is needed.

CMCDragonkai replied:
Which issue is this for? I thought we had discussed earlier and it already been fixed.

CMCDragonkai replied:
If this is in fact a proper fix, please create an issue and fix it with this PR, and give demonstration of what exactly this fixes. Like if it is to reduce logger spam. Then post an example of what is expected or not expected during execution by copying the terminal output.

tegefaulkes replied:
Ok, I'll do that.

CMCDragonkai commented on src/websockets/WebSocketServer.ts:
```diff
@@ -96,7 +96,7 @@ class WebSocketServer extends EventTarget {
   protected server: uWebsocket.TemplatedApp;
   protected listenSocket: uWebsocket.us_listen_socket;
-  protected host: string;
+  protected host_: string;
```
Change to `this._host`.

tegefaulkes replied:
Updated.

CMCDragonkai replied:
Check if anything else is still `x_` and not `_x`.

CMCDragonkai commented on tests/client/service/agentLockAll.test.ts:
Why are all service tests deleted?

tegefaulkes replied:
Those are the old tests, the new ones are under `tests/client/handlers/...`

CMCDragonkai reviewed (changes requested):
Initial review.

tegefaulkes committed:
fix: updated error names, `Rpc` is now `RPC`

[ci skip]

tegefaulkes committed:
fix: added return types to utils functions

[ci skip]

tegefaulkes committed:
syntax: renamed `host_` to `_host` to match conventions

[ci skip]

tegefaulkes committed:
fix: adding docblock documentation to middleware.ts

[ci skip]

tegefaulkes committed:
fix: remove duplicate error

[ci skip]

tegefaulkes committed:
fix: making `PolykeyClient` transport agnostic

[ci skip]

tegefaulkes committed:
fix: changing order of discovery message spam fix

* Related #514

[ci skip]

tegefaulkes committed:
fix: renaming `streamPairCreateCallback` to `streamFactory`

[ci skip]

tegefaulkes committed:
fix: using transport agnostic `PolykeyClient`

[ci skip]

tegefaulkes committed:
fix: removed placeholder error and fixed error message handling

[ci skip]

tegefaulkes committed:
syntax: renamed metadata `Authorization` to `authorization`

[ci skip]

tegefaulkes committed:
syntax: updated client response and request data types

[ci skip]

tegefaulkes committed:
fix: fixing up directory structures of `client` and `rpc`

[ci skip]

CMCDragonkai commented on src/rpc/errors.ts:
```diff
@@ -0,0 +1,107 @@
+  ErrorRPCMessageLength,
+  ErrorRPCMissingResponse,
+  ErrorPolykeyRemote,
+  ErrorSendErrorFailed,
```
This should be `ErrorRPCSendErrorFailed`.

CMCDragonkai commented on src/rpc/errors.ts:
```diff
@@ -0,0 +1,107 @@
+  exitCode = sysexits.UNAVAILABLE;
+}
+
+class ErrorPolykeyRemote<T> extends ErrorPolykey<T> {
```
I'm moving this to the bottom. Just like in `grpc/errors`.

CMCDragonkai committed:
fix(rpc): moved `ErrorPolykeyRemote` to the bottom of `rpc/errors` and changed `ErrorSendErrorFailed` to `ErrorRPCSendErrorFailed`

CMCDragonkai commented on src/rpc/RPCServer.ts:
```diff
@@ -217,22 +217,26 @@ class RPCServer {
-                    error: e,
+                    error: new rpcErrors.ErrorRPCSendErrorFailed(undefined, {
+                      cause: [e, reason],
+                    }),
```
Is this expected to bubble up on the agent side, or does this get sent to the remote side?

tegefaulkes replied:
This is purely the server side. It wraps the errors that happen when the stream throws an error. So it's only emitted as an error event for the server.

CMCDragonkai replied:
Hmm want to rename this a bit.

CMCDragonkai replied:
See the discussion: https://github.com/MatrixAI/Polykey/pull/509#issuecomment-1480565073

CMCDragonkai commented on src/rpc/RPCServer.ts:
```diff
@@ -202,7 +202,7 @@ class RPCServer {
             controller.enqueue(value);
           } catch (e) {
-            if (rpcUtils.isReturnableError(e)) {
+            if (reason == null) {
```
What was the reason behind this? That is changing from `isRetunrable` to `null`?

tegefaulkes replied:
After looking over the placeholder error I determined that it wasn't needed and the current method didn't really work. Currently the only time we want to not send an error through the stream is when we can't. So now we're only checking if the stream has already failed. if so we emit the error through the event system otherwise we just send the error message.

CMCDragonkai replied:
How come you don't check `e`? You're checking `reason` here.

tegefaulkes replied:
Reason is set by the streams `abort` handler. It lets us know that we can't enqueue the error message if we wanted to. It `reason` could use a more descriptive name in that case.

CMCDragonkai replied:
What would `e` be if `reason` is `null`?

tegefaulkes replied:
`e` would be an error thrown in the handler if it failed to do something.

CMCDragonkai replied:
> Reason is set by the streams `abort` handler. It lets us know that we can't enqueue the error message if we wanted to. It `reason` could use a more descriptive name in that case.

You mean the `cancel` handler?

tegefaulkes replied:
Yes.

CMCDragonkai replied:
See discussion https://github.com/MatrixAI/Polykey/pull/509#issuecomment-1480552627

CMCDragonkai committed:
fix: removed spurious mention of `ErrorPolykeyRemote`

CMCDragonkai commented on src/PolykeyAgent.ts:
```diff
@@ -103,7 +107,6 @@ class PolykeyAgent {
     vaultManager,
     notificationsManager,
     sessionManager,
-    grpcServerClient,
```
Don't just remove this and then establish it in the agent. Preserve the pattern that we are using. Dependency inject the `rpcServerClient`. Or whatever is relevant here.

Plus you didn't update the types to remove `grpcServerClient`, that's still there.

The original thing here was meant to be the RPC server for the client service. That's why there's also the `grpcServerAgent`.

I'm guessing now when we migration the agent service to to this, would we also have a `rpcServerAgent`?

Furthermore you would have 2 RPC servers, one for the client service and one for the agent service.

And I also think the `webSocketServer` should also be DIed like the rest of it.

CMCDragonkai replied:
Note that:

```
    grpcServerClient?: GRPCServer;
```

Still exists in the types here.

tegefaulkes replied:
I missed the `grpcServerClient`, I'll remove it.

As for DI, I can make them both DI-ed.

tegefaulkes replied:
I don't think the rpcServer can be DI-ed. It has a mutual reference to the `PKAgent`.  This means that it needs to be created within the `PolykeyAgent`'s constructor. '

It's cheesy but we can pass the `PKAgent` into it as a promise to be resolved later. that might be the better option here. I think the only problem is, if you don't resolve the promise, the agent stop handler will hang until timeout.

tegefaulkes replied:
I went with the promise method. Seems like the easiest solution to the problem.

tegefaulkes replied:
This has been fixed, `rpcServerClient` and `webSocketServerClient` can now be dependency injected.

CMCDragonkai commented on src/PolykeyAgent.ts:
```diff
@@ -401,11 +405,15 @@ class PolykeyAgent {
+        undefined,
+        1_000,
+        10_000,
+      );
```
This should be DIed like `webSocketServer = webSocketServer ?? new WebSocketServer(...)`.

Also what do you mean by propagating the default values?

tegefaulkes replied:
There are config values that are just hard-coded here. They need to be added to the `PolykeyAgent` config and defaults.

tegefaulkes replied:
This has been changed.

CMCDragonkai commented on src/PolykeyClient.ts:
```diff
@@ -57,20 +54,17 @@ class PolykeyClient {
+        authMiddleware.authenticationMiddlewareClient(session),
+      ),
+      logger: logger.getChild(RPCClient.name),
+    });
```
Change this to be DIed optionally like we did with `grpcClient`. The DI enables mock testing.

tegefaulkes replied:
I'm thinking about how to set the manifest or DI the `RPCClient`. Technically the options are mutually exclusive. Ideally this would be enforced with the type.

I can think of two options. 
1. the `manifest` and `rpcClient` are both optional but creation will fail if if neither are provided. Providing both would default to the DI-ed `rpcClient`.
2. The other option is to have the `manifest` be the `RPCClient|ClientManifest`, if the manifest is provided then we create the `rpcClient`. 

I'm not sure which one is preferred here.

tegefaulkes replied:
I'm going with option 2, it's much cleaner.

tegefaulkes replied:
This is fixed.

CMCDragonkai commented on src/PolykeyClient.ts:
```diff
@@ -114,13 +108,13 @@ class PolykeyClient {
     await this.session.stop();
     this.logger.info(`Stopped ${this.constructor.name}`);
   }
 
   public async destroy() {
     this.logger.info(`Destroying ${this.constructor.name}`);
+    await this.rpcClient.destroy();
```
So we moved from stopping to destroying for the `RPCClient`?

tegefaulkes replied:
Yeah, the `RPCClient` doesn't really have any state anymore. It's just a container for some data and an interface for making calls. It doesn't really need to be destroyed even. it only prevents it from being used from then on.

CMCDragonkai commented on src/bin/agent/CommandLockAll.ts:
```diff
@@ -16,9 +17,11 @@ class CommandLockAll extends CommandPolykey {
       const { default: Session } = await import('../../sessions/Session');
-      const utilsPB = await import('../../proto/js/polykey/v1/utils/utils_pb');
-
+      const { clientManifest } = await import('../../client/handlers');
```
This imports EVERYTHING! The `client/handlers/index.ts` is importing the entire project.

Don't do this. Remember the whole reason we are doing this is `await import` is to specialise to the only the code necessary for the command.

You're making every command dynamically load the entire project source with `await import('../../client/handlers');`

You need to have a more specific import.

The original thing was just `utilsPB` which was just a simple structure for utilities.

Now this ends up slowing the command execution.

I see the way you're creating the `clientManifest`. I think it's too heavy, either we need a different way, or can we do without having a manifest, or can we only import **exactly** what we need without going through the `index.ts` imports.

tegefaulkes replied:
The core of this is we need a way to define the manifest without importing everything. The reason everything is being imported because the client caller definitions are defined in the same file of the handler.  So we're implicitly importing the handler dependencies.

The only thing the callers need are the types involved. but we need to keep them in a logical place close to the handlers because they need to be created and updated side by side. The reason for this is that the client and agent definitions are closely coupled by type, not enforced by any type system and if they don't match exactly we end up with type bugs that are hard to debug without intimate knowledge of the system. 

As for a fix for this. We need to define the server and client manifest in seperate files and we can't defined the callers in the handlers file. I need to make a `clientManifest.ts` file and possibly a server variant for this. I'm not sure if these files should be kept in the `handlers` directory or the `client` directory. 

I'll keep them in the handlers directory for now.

tegefaulkes replied:
The `clientManifest` and `serverManifest` are now in their own respective files, `clientManifest.ts` and `serverManifest.ts`. They are both still exported by the index. 

We are now importing just the `clientManifest` for use with the `rpcClient`. This should only depend on type imports now. 

To get any more specific than this then we'd need separate files for each caller definition. That would be too much to maintain for little benifit.

CMCDragonkai commented on src/bin/agent/CommandLockAll.ts:
```diff
@@ -39,27 +42,37 @@ class CommandLockAll extends CommandPolykey {
+        pkClient = await PolykeyClient.createPolykeyClient({
+          streamFactory: () => webSocketClient.startConnection(),
           nodePath: options.nodePath,
+          manifest: clientManifest,
```
There should be a way of acquiring this information without importing the entire application. I mean also in a transitive sense.

tegefaulkes replied:
For the bin commands we can just define the types for the RPC methods we're using in the command. The problem is, it's essential that they match otherwise very annoying to debug bugs happen. To mitigate this the caller and handler are defined side by side in the same location. But as you said this leaves us depending on a bunch of stuff we don't need. 

We can just define the caller in the command and use a truncated manifest, Or separate the client manifest definition to a new file so it only depends on the types it needs.

tegefaulkes replied:
This is the same problem as the above comment. I'll do the discussion there.

CMCDragonkai commented on src/bin/polykey-agent.ts:
```diff
@@ -114,8 +114,8 @@ async function main(_argv = process.argv): Promise<number> {
-    clientHost: pkAgent.grpcServerClient.getHost(),
-    clientPort: pkAgent.grpcServerClient.getPort(),
+    clientHost: pkAgent.webSocketServer.host,
+    clientPort: pkAgent.webSocketServer.port,
```
I'm trying to find where `getHost` and `getPort` methods used to exist. I don't see them on `GRPCClient`. There would be a reason why these getters exist, usually when the property must be mutable, but only internally so `readonly` modifier cannot be used.

tegefaulkes replied:
They weren't getters before, just normal methods. I made them getters since the host and port is updated every time you start. A `readonly` can only be set during construction so I couldn't update it every time we started. 

So we have a getter here. The only alternative is to change it to match the`getHost()` style.

CMCDragonkai replied:
I think of `getHost` as getters. The `get host` is not a getter. It's a magic method overloading the `.host`.

tegefaulkes replied:
Sure, but the technical term for the magic methods overriding `.host` are getters and setters aren't they?

Is there a change to be made here?

CMCDragonkai replied:
The reason to use `.getHost()`...

```ts
class X {
  // this means x is readable and immutable
  readonly x;
  // this means y is readable and mutable
  y;
  _z
  // this means z is readable, but internally mutable, but externally immutable
  get z () {
    return this._z;
  }
  // why use getA over the above?
  // supposed to mean that the A property is initialised after construction, or has specialised construction
  getA() {

  }
}
```

The `host` and `port` are not directly set before construction, they occur after an asynchronous start. That's why I always make them `getHost` and `getPort` to distinguish them from static properties. We've also talked about this distinction before. Then `@ready` can be applied too here, where one can say that `host` and `port` are only available after the server is started. Prior to this, you may do something like `await Server.createServer({ host: '0.0.0.0', port: 0 })`. But then afterwards it is "resolved", with `server.getPort() // returns 1235`.

The point is `host` and `port` are dynamic properties, and I reserve `.a` or `.b` properties to be static properties.

tegefaulkes replied:
Ok, This has been fixed.

CMCDragonkai commented on src/bin/utils/processors.ts:
```diff
@@ -11,10 +11,9 @@ import type {
 import Logger from '@matrixai/logger';
 import Status from '../../status/Status';
-import * as clientUtils from '../../client/utils';
+import * as clientUtils from '../../client/utils/utils';
```
This isn't actually what I meant. In many other situations what we do when we have this kind of code is to do this:

```
utils/index.ts
utils/x.ts
utils/y.ts
```

You don't want to do `utils/utils.ts`. That defeats the purpose of putting it into a directory. It's to have file-level separation which can be more efficient for importing or navigation and code review.

CMCDragonkai replied:
Indeed there is another one with:

```
utils.ts
utils/x.ts
```

I don't like this as it is confusing, many other languages have confusing overlaps here, it's better to use `utils/index.ts`.

You can have broad root-level utilities in `utils/index.ts`, and you can specialise certain utilities into other files.

CMCDragonkai replied:
Ok there is an example with `src/bin/utils/utils.ts`.

The idea being that `src/bin/utils/index.ts` does a flat export of `src/bin/utils/utils.ts`.

But the idea of specialisation is still important to have separate files.

tegefaulkes replied:
Ok, so you're saying that the methods in `utils/utils.ts` should be moved into the `utils/index.ts`?

CMCDragonkai replied:
Not necessarily, it's about splitting up the methods into files. I'm looking into the `rpc` now to see what factorisation can occur.

tegefaulkes replied:
I'm confused then. I created the utils directory in the client and `rpc` domain to store the `utils.ts` and `middleware.ts` files together. I can maybe split the `utils.ts` files up more? haven't I done what you are saying here?

CMCDragonkai replied:
Will leave this to later, this requires analysis of the middleware functions and how best to organise them.

CMCDragonkai commented on src/bin/utils/utils.ts:
```diff
@@ -112,7 +112,7 @@ function outputFormatter(msg: OutputObject): string | Uint8Array {
     let indent = '  ';
     while (currError != null) {
-      if (currError instanceof grpcErrors.ErrorPolykeyRemote) {
+      if (currError instanceof grpcErrors.ErrorPolykeyRemoteOLD) {
```
Don't we need to have something like this but for the new JSON based RPC?

tegefaulkes replied:
We do have it. it's called `ErrorPolykeyRemote`. I had to rename the old one to avoid conflict until it's fully removed.

CMCDragonkai replied:
I know I meant the line itself. You're checking for the remote old, why not check for the `ErrorPolykeyRemote` too?

tegefaulkes replied:
Ah I see, I misunderstood. It should be checking for both until the old one is removed. I'll fix this up.

CMCDragonkai commented on src/client/errors.ts:
```diff
@@ -1,30 +1,24 @@
-  ErrorClient,
-  ErrorClientClientDestroyed,
+  ErrorRPCClient,
   ErrorClientAuthMissing,
   ErrorClientAuthFormat,
   ErrorClientAuthDenied,
```
All these should the be `ErrorRPCClientX`. Always keep to consistent naming.

On the other hand, originally I just called it `ErrorClient` to avoid the `RPC`. You can also change `ErrorRPCClient` to just `ErrorClient`.

CMCDragonkai replied:
This meaning it is an error inside the client service.

tegefaulkes replied:
I don't think so? They are different domains. These are the Client errors for the `client` domain., the other ones are `rpc` domain errors for the RPCClient. They are completely different.

CMCDragonkai replied:
Yea so then `ErrorRPCClient` should be `ErrorClient`?

CMCDragonkai replied:
You're extending `ErrorRPCClient` atm, and the subtypes are just `ErrorClientX`. So instead it should be `ErrorClientX` extending `ErrorClient`.

tegefaulkes replied:
Ah, I didn't see that in the context of the comment. I'll fix it.

tegefaulkes committed:
fix: `rpcServerClient` and `webSocketServerClient` are not dependency injected into the `PolykeyAgent`.

[ci skip]

tegefaulkes committed:
fix: `rpcClient` is now dependency injected into `PolykeyClient`

[ci skip]

CMCDragonkai commented on src/rpc/RPCServer.ts:
```diff
@@ -27,9 +26,9 @@ import {
+import * as middlewareUtils from './utils/middleware';
 import { never } from '../utils/utils';
 import { sysexits } from '../errors';
 
```
We need some docblock for `sensitive`. Well docblock in `createRPCServer` and explain what does `sensitive` mean.

tegefaulkes replied:
I've added docblocks to both `RPCClient` and `RPCServer` filling in some details.

CMCDragonkai commented on src/rpc/types.ts:
```diff
@@ -228,6 +229,13 @@ type MapCallers<T extends ClientManifest> = {
+
 export type {
   JsonRpcRequestMessage,
   JsonRpcRequestNotification,
```
I'm changing all of this to be `JSONRPC...`.

CMCDragonkai committed:
fix: changed `JsonRpc` to `JSONRPC`

CMCDragonkai commented on src/rpc/RPCServer.ts:
```diff
@@ -44,10 +43,10 @@ class RPCServer {
   }: {
     manifest: ServerManifest;
     middleware?: MiddlewareFactory<
```
Why is `middleware` a `MiddlewareFactory`? I think it should be called `middlewareFactory` since you later call it.

Also please add a docblock to the `middleware` property explaining its types and how it is intended to work.

tegefaulkes replied:
Done.

CMCDragonkai commented on src/rpc/RPCServer.ts:
Are all the `manifestItem.handle` methods or can they be arrow functions? If they are arrow functions, there's no need to `bind` them.

tegefaulkes replied:
They need to access this for the context of the handler. So I don't think they can be arrow functions here.

CMCDragonkai replied:
The `this` context can exist in arrow function properties.

```ts
class C {
  f = () => {
    console.log(this);
  }
}
```

CMCDragonkai replied:
But I can see you made each `handle` an abstract method. I wonder if it could be done as an abstract property.

CMCDragonkai replied:
It actually does work if you do this:

```
  abstract handle: (
    input: AsyncIterable<Input>,
    connectionInfo: ConnectionInfo,
    ctx: ContextCancellable,
  ) => Promise<Output>;
```

That's an abstract arrow function property.

We can deal with this later.

CMCDragonkai replied:
Can create new issue for this.

tegefaulkes replied:
Created.

CMCDragonkai commented on src/rpc/types.ts:
```diff
@@ -12,11 +12,12 @@ import type {
  */
-type JsonRpcRequestMessage<T extends JSONValue = JSONValue> = {
+type JSONRPCRequestMessage<T extends JSONValue = JSONValue> = {
   // A String specifying the version of the JSON-RPC protocol. MUST be exactly "2.0"
```
I think if you want these comments to show up in the typedoc generated comments, all of these must be docblocks, not `//` comments.

We only use line comments for inside functions/procedures.

tegefaulkes replied:
They're mostly for reference and explaining what each field is for. It doesn't need to show up in the IDE docs.

CMCDragonkai replied:
Not the IDE docs, the typedocs. If it's switched to docblocks, you should be able to see it too.

tegefaulkes replied:
I've made them doc comments and some to the message types.

CMCDragonkai commented on src/rpc/types.ts:
```diff
@@ -84,7 +85,7 @@ type JsonRpcResponseError = {
 //  -32603	Internal error	Internal JSON-RPC error.
 //  -32000 to -32099
```
I don't think we should be using these codes.

Instead you should just have unique exception classes for each possible error.

The reason is because in an OOP language like ourselves, the class name itself is sufficient identification, code is unnecessary.

We do use `code` for sysexit codes, but that's only because a number is necessary when exiting a program.

When sending the error over the RPC, we don't need it, we only need the serialised version of the exception object which is then later deserialised on the remote side.

CMCDragonkai replied:
So you should just literally have:

```
class ErrorRPCParseError { }
class ErrorRPCInvalidRequest { }
class ErrorRPCMethodNotFound { }
class ErrorRPCInvalidParams { }
class ErrorRPCInternalError { }
```

The name itself is unique and more descriptive than an arbitrary number.

CMCDragonkai replied:
If the JSON RPC spec requires a `code: number`, we can fill it with our sysexit code. But the really we should be replacing that with our "serialised" error structure. The one we use with `fromJSON` and `toJSON`.

tegefaulkes replied:
We're not using these codes anywhere. They were for reference. They can be removed.

CMCDragonkai committed:
style: spacing between handler types

CMCDragonkai commented on src/rpc/RPCServer.ts:
When you do:

```ts
  protected registerUnaryHandler<I extends JSONValue, O extends JSONValue>(
    method: string,
    handler: UnaryHandlerImplementation<I, O>,
  ) {
    const wrapperDuplex: DuplexHandlerImplementation<I, O> = async function* (
      input,
      connectionInfo,
      ctx,
    ) {
      for await (const inputVal of input) {
        yield handler(inputVal, connectionInfo, ctx);
        break;
      }
    };
    this.registerDuplexStreamHandler(method, wrapperDuplex);
  }
```

The `yield handler(...)`, should there be an await there? Like `yield await handler(...)`?

CMCDragonkai replied:
The whole point of having an async generator function is so that you can use `await` inside. But the `handler` here returns a promise. So why not await the promise and yield back the result?

tegefaulkes replied:
It's not strictly needed, the only difference is how the call stack looks if any error was thrown. The handler will be awaited when awaiting the value from the `AsyncGenerator`.  There is no specific reason it was done this way, I can add an await in.

CMCDragonkai replied:
Yea I think ti should be changed to `await`. That seems like the most common way of doing it.

CMCDragonkai replied:
Plus it is better for debugging purposes here.

I can understand using such things for function wrappers. But in these situations, it's better safe than sorry.

CMCDragonkai commented on src/rpc/RPCServer.ts:
```diff
@@ -357,19 +360,19 @@ class RPCServer {
 
-  @ready(new rpcErrors.ErrorRpcDestroyed())
+  @ready(new rpcErrors.ErrorRPCDestroyed())
   public removeEventListener(
```
Question... why didn't you just extend the `EventTarget`?

tegefaulkes replied:
I can, but i'd lose the type information for the handler. dealing with the raw event target with custom events with added data can be confusing and annoying.  This also allows for the `@ready` decorator which I'm not sure is strictly needed.

CMCDragonkai replied:
I'm extending it in the QUIC code base. I'm sure it's possible to do this without it. Also possible to have an overloaded signature in the subclass I think. I'll try.

CMCDragonkai replied:
I've changed to extension here and also simplified the event class itself. Alignment with js-quic is more important here. However subclasses cannot provide a more specialised signature unfortunately. Maybe there's a hack, but not important.

tegefaulkes committed:
fix: `clientManifest` avoids needless dependencies

Previously the `clientManifest` depended on the whole code base due to the handlers. This has now been seperated out, so it only depends on types.

[ci skip]

tegefaulkes committed:
fix: `outputFormatter` handles `ErrorPolykeyRemote` as well as the old one for now

[ci skip]

tegefaulkes committed:
fix: renamed `ErrorRPCClient` to `ErrorClient`

[ci skip]

tegefaulkes committed:
fix: removing a disabled test

[ci skip]

tegefaulkes committed:
fix: updated docblocks and parameter names for `RPCServer` and `RPCClient`

[ci skip]

CMCDragonkai committed:
chore: removed useless comments about XML RPC error codes

CMCDragonkai committed:
fix(rpc): await the handler promises in the async generators, this is better for debugging

CMCDragonkai committed:
fix(rpc): moved `RPCErrorEvent` to `rpc/events`, collapsed `Error` to `detail` property, made `RPCServer` extend `EventTarget`

CMCDragonkai committed:
chore: expanding `RPCServer` commentary

CMCDragonkai commented on src/rpc/RPCServer.ts:
```diff
@@ -253,7 +282,7 @@ class RPCServer {
       ctx,
     ) {
       for await (const inputVal of input) {
-        yield handler(inputVal, connectionInfo, ctx);
+        yield await handler(inputVal, connectionInfo, ctx);
         break;
       }
```
Given that this is a unary handler, why are we doing a `for await` on the `input`?

Shouldn't we expect just one piece of data from the `input`?

That is instead of doing `for await`, it would be clearer to do:

```
const inputVal = input.next();
```

---

Oh wait, because it is an `AsyncIterable`, you have to use `for await`.

I am adding a comment here.

CMCDragonkai replied:
It is possible to do this:

```
const inputVal = (await input[Symbol.asyncIterator]().next()).value;
```

However it's a bit wordy.

tegefaulkes committed:
fix: made `WebSocketServer` `host` and `port` into `getHost()` and `getPort()` and other fixes

[ci skip]

tegefaulkes committed:
chore: updating doc comments for RPC message types

[ci skip]

CMCDragonkai committed:
fix: change of head, tail destructuring for raw handlers, head and tail reconstruction via transform stream, and reverse pair propagation of cancellation event

tegefaulkes committed:
fix: `RPCServer` handling output stream cancellation

[ci skip]

tegefaulkes committed:
fix: cleaning up temp errors in websocket domain

[ci skip]

tegefaulkes committed:
fix: handler receives signal if output stream fails

Enabling CI.
